src/libcollections/vec.rs

```diff
@@ -1652,16 +1521,16 @@ impl<T: Ord> Ord for Vec<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Drop for Vec<T> {
     fn drop(&mut self) {
-        // This is (and should always remain) a no-op if the fields are
-        // zeroed (when moving out, because of #[unsafe_no_drop_flag]).
-        if self.cap != 0 && self.cap != mem::POST_DROP_USIZE {
-            unsafe {
-                for x in self.iter() {
-                    ptr::read(x);
-                }
-                dealloc(*self.ptr, self.cap)
+        // NOTE: this is currently abusing the fact that ZSTs can't impl Drop.
+        // Or rather, that impl'ing Drop makes them not zero-sized. This is
+        // OK because exactly when this stops being a valid assumption, we
+        // don't need unsafe_no_drop_flag shenanigans anymore.
+        if mem::size_of::<T>() != 0 && self.buf.cap() != mem::POST_DROP_USIZE {
```

Comment:
Discussed this on IRC, but to re-iterate what I said: I don't consider unsafe_no_drop_flag ZSTs to be sound. If you use them you're wrong and Vec is allowed to be arbitrarily malicious with them. CC @pnkfelix